### PR TITLE
Change function kernels to use CustomOp APIs

### PR DESCRIPTION
**Pull request by RyanUnderhill**

### CODE REVIEW

File: include/onnxruntime/core/framework/func_api.h [hunk 1]
```diff
@@ -1,31 +1,6 @@
 #pragma once
 #include "core/common/common.h"
 namespace onnxruntime {
-//TODO: Should use the lotus cpi element type definition.
-enum DType {
```
Comment (snnn) [inline include/onnxruntime/core/framework/func_api.h]: Hey, I have a similar change in another PR.  #989
As yours is smaller, I'd prefer to let this one go first.

**Comment by jywu-msft** [comment]
thanks for the refactoring. since the api is changed now, TensorRT and nGraph providers need to be updated (especially TensorRT which is being used in 1P production)

**Comment by jywu-msft** [comment]
also nuphar

> thanks for the refactoring. since the api is changed now, TensorRT and nGraph providers need to be updated (especially TensorRT which is being used in 1P production)

**Comment by KeDengMS** [comment]
> also nuphar
> 
> thanks for the refactoring. since the api is changed now, TensorRT and nGraph providers need to be updated (especially TensorRT which is being used in 1P production)

Thanks for the notice. Nuphar will be updated after this is merged.

**Comment by souptc** [comment]
LGTM, thanks Ryan for this refactor, we finally get rid of the ugly tensor copy in function kernel.

**Pull request merged**